src/TestSuite/Fixture/FixtureHelper.php

```diff
@@ -159,11 +171,71 @@ public function insert(array $fixtures): void
     public function truncate(array $fixtures): void
     {
         $this->runPerConnection(function (ConnectionInterface $connection, array $groupFixtures) {
-            $connection->disableConstraints(function () use ($connection, $groupFixtures) {
-                foreach ($groupFixtures as $fixture) {
+            if ($connection instanceof Connection) {
+                $sortedFixtures = $this->sortByConstraint($connection, $groupFixtures);
+            }
+
+            $driver = $connection->getDriver();
+            if (
+                isset($sortedFixtures) &&
+                (
+                    $driver instanceof Postgres ||
+                    $driver instanceof Sqlite ||
+                    $driver instanceof Sqlserver
+                )
+            ) {
+                foreach (array_reverse($sortedFixtures) as $fixture) {
                     $fixture->truncate($connection);
                 }
-            });
+            } else {
+                $connection->disableConstraints(function () use ($connection, $groupFixtures) {
+                    foreach ($groupFixtures as $fixture) {
+                        $fixture->truncate($connection);
+                    }
+                });
+            }
         }, $fixtures);
     }
+
+    /**
+     * Sort fixtures with foreign constraints last if possible, otherwise returns null.
+     *
+     * @param \Cake\Database\Connection $connection Database connection
+     * @param array<\Cake\Datasource\FixtureInterface> $fixtures Database fixtures
+     * @return array|null
+     */
+    protected function sortByConstraint(Connection $connection, array $fixtures): ?array
+    {
+        static $schemas = [];
+
+        $fixturesWithConstraint = [];
+        $fixturesWithoutConstraint = [];
+        foreach ($fixtures as $fixture) {
+            $tableName = $fixture->sourceName();
+            if (!isset($schemas[$tableName])) {
+                $schemas[$tableName] = $connection->getSchemaCollection()->describe($tableName);
+            }
+
+            $schema = $schemas[$tableName];
+            foreach ($schema->constraints() as $constraintName) {
+                $constraint = $schema->getConstraint($constraintName);
+                if (!$constraint) {
+                    continue;
+                }
+
+                if ($constraint['type'] === TableSchema::CONSTRAINT_FOREIGN) {
+                    $fixturesWithConstraint[$schema->name()] = $fixture;
+                    if (isset($fixturesWithConstraint[$constraint['references'][0]])) {
+                        return null;
+                    }
+                }
+            }
+
+            if (!isset($fixturesWithConstraint[$schema->name()])) {
+                $fixturesWithoutConstraint[$schema->name()] = $fixture;
+            }
+        }
+
+        return array_merge(array_values($fixturesWithoutConstraint), array_values($fixturesWithConstraint));
```

Comment:
Couldn't we end up with fixtures in an order that we can't truncate because parent tables come are sorted after the children? Or are we relying on users to define fixtures in the required order? I am ok with people needing to put their fixtures in the correct order to satisfy constraints.  If we have that requirement  (which I think we have always had) do we need to sort fixtures?

Reply:
They order is revered in truncate.